### PR TITLE
tms34020: Implement SETCDP opcode

### PR DESCRIPTION
**Pull request by plaes**
This pull request implements `SETCDP` opcode for tms34020.

Standalone testcase here: https://gist.github.com/plaes/dec4b76ffc4b720726f21031d8d399ac

Changes in v3:
* Dropped log2 usage in favor of `32 - count_leading_zeros()`
* Dropped useless includes
* Fixed formatting for braces
* Use logical or instead of sum

Changes in v2:
* Use functions from eminline.h (population_count_32 instead of __builtin_popcount and count_leading_zeros instead of __builtin_ffs)
* Use `log2` from `<cmath>`
* Fix pointless `log2` calls in the double power of two case.

### CODE REVIEW

**Comment by galibert** [comment]
Please use popcount32/popcount64 and count_leading_zereos/ones from
emuinline.h instead.

  OG.

File: src/devices/cpu/tms34010/34010ops.hxx [hunk 1]
```diff
@@ -2392,10 +2392,51 @@ void tms340x0_device::rpix_b(uint16_t op)
...
+				COUNT_CYCLES(4);
+				return;
+			}
+			// .. two bits, we can decompose it to sum of two power of two numbers
+			case 2:
+			{
+				uint8_t b1 = __builtin_ffs(dptch) - 1;
+				uint8_t b2 = __builtin_ffs(dptch & ~(1 << b1)) - 1;
+
+				uint32_t val1 = log2(1 << b1);
```
Comment (cuavas) [inline src/devices/cpu/tms34010/34010ops.hxx]: The base 2 logarithm of `1 << x` is `x`.  Think about it: `1 << x` is the same operation as two to the power of `x` (let's call it `2 ** x`), so `log2(1 << x)` is equivalent to `log2(2 ** x)`, which is equal to `x`.  You definitely need to re-think your maths here.

Comment (plaes) [reply src/devices/cpu/tms34010/34010ops.hxx]: Thanks, I had misread the manual. Testcase with 4 sample from tms34020 manual is here:

https://gist.github.com/plaes/dec4b76ffc4b720726f21031d8d399ac

**Comment by plaes** [comment]
I pushed updated implementation. Please check the top comment for changes.

File: src/devices/cpu/tms34010/34010ops.hxx [hunk 2]
```diff
@@ -2395,7 +2399,40 @@ void tms340x0_device::rpix_b(uint16_t op)
...
+	// logerror("020:setcdp: PC=%x, DPTCH()=%x\n", m_pc, dptch);
+
+	// Check whether we're dealing with an even number
+	if ((dptch & 1) == 0) {
+		switch(population_count_32(dptch))
+		{
+			// .. only single bit set, pitch is power of two!
+			case 1:
+			{
+				uint32_t val = log2(dptch);
```
Comment (cuavas) [inline src/devices/cpu/tms34010/34010ops.hxx]: You can get this from counting leading zeroes.  You don't want to invoke floating-point maths.

File: src/devices/cpu/tms34010/34010ops.hxx [hunk 3]
```diff
@@ -2395,7 +2399,40 @@ void tms340x0_device::rpix_b(uint16_t op)
...
 	if (!m_is_34020) { unimpl(op); return; }
-	logerror("020:setcdp\n");
+	off_t dptch = DPTCH();
+
+	// logerror("020:setcdp: PC=%x, DPTCH()=%x\n", m_pc, dptch);
+
+	// Check whether we're dealing with an even number
+	if ((dptch & 1) == 0) {
+		switch(population_count_32(dptch))
+		{
```
Comment (cuavas) [inline src/devices/cpu/tms34010/34010ops.hxx]: Be consistent with code style - if the file uses K&R style, don't use Allman, and vice versa.

File: src/devices/cpu/tms34010/34010ops.hxx [hunk 4]
```diff
@@ -2395,7 +2399,40 @@ void tms340x0_device::rpix_b(uint16_t op)
...
+				return;
+			}
+			// .. two bits, we can decompose it to sum of two power of two numbers
+			case 2:
+			{
+				uint8_t val1 = count_leading_zeros(dptch) - 1;
+				uint8_t val2 = count_leading_zeros(dptch & ~(1 << val1)) - 1;
+
+				val1 = ~val1 & 0x1f;
+				val2 = ~val2 & 0x1f;
```
Comment (cuavas) [inline src/devices/cpu/tms34010/34010ops.hxx]: Isn't this equivalent to `31 - x`?  It's going to be easier for the compiler to optimise it if you do this as a subtraction integrated into the operation above.

File: src/devices/cpu/tms34010/34010ops.hxx [hunk 5]
```diff
@@ -2395,7 +2399,40 @@ void tms340x0_device::rpix_b(uint16_t op)
...
+
+				m_convdp = val2 + (val1 << 8);
```
Comment (cuavas) [inline src/devices/cpu/tms34010/34010ops.hxx]: The values shouldn't overlap, so it's better to express this as a logical or, not an addition.

File: src/devices/cpu/tms34010/34010ops.hxx [hunk 6]
```diff
@@ -10,6 +10,10 @@
 ***************************************************************************/
 
 
+// for log2
+#include <cmath>
+// for popcount32, count_leading_zeroes
+#include "eminline.h"
```
Comment (cuavas) [inline src/devices/cpu/tms34010/34010ops.hxx]: This should already be included via `emu.h`

**Comment by happppp** [comment]
"Fix pointless log2 calls in the double power of two case." But it's still there, did you forget to update the file? Case 1 can be reduced to:
m_convdp = count_leading_zeros(dptch);
Here, log2(x) is equal to: 31-count_leading_zeros(x), let's call this "y". And ~(31-y) & 0x1f is simply y.

You don't need to include eminline, it is already included by default since the file includes emu.h

File: src/devices/cpu/tms34010/34010ops.hxx (hunk 6 above)
Comment (plaes) [reply src/devices/cpu/tms34010/34010ops.hxx]: OK, dropped both includes.

**Comment by plaes** [comment]
Hopefully resolved all the issues, and rebased.

**Pull request merged**